core/imt/inc/ROOT/TExecutor.hxx

```diff
@@ -0,0 +1,337 @@
+// @(#)root/thread:$Id$
+// Author: Xavier Valls September 2020
+
+/*************************************************************************
+ * Copyright (C) 1995-2006, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+#ifndef ROOT_TExecutor
+#define ROOT_TExecutor
+
+#include "ROOT/TExecutorBaseImpl.hxx"
+#include "ROOT/TSequentialExecutor.hxx"
+#ifdef R__USE_IMT
+#include "ROOT/TThreadExecutor.hxx"
+#endif
+#include "ROOT/TProcessExecutor.hxx"
+#include "TROOT.h"
+#include "ExecutionPolicy.hxx"
+#include <memory>
+#include <stdexcept>
+#include <thread>
+
+namespace ROOT{
+
+namespace Internal{
+class TExecutor: public TExecutorBaseImpl<TExecutor> {
+public:
+
+   explicit TExecutor(unsigned nProcessingUnits = 0) :
+      TExecutor(ROOT::IsImplicitMTEnabled() ? ROOT::Internal::ExecutionPolicy::kMultithread : ROOT::Internal::ExecutionPolicy::kSerial, nProcessingUnits) {}
+
+   explicit TExecutor(ROOT::Internal::ExecutionPolicy execPolicy, unsigned nProcessingUnits = 0) : fExecPolicy(execPolicy) {
+      fExecPolicy = execPolicy;
+      switch(fExecPolicy) {
+         case ROOT::Internal::ExecutionPolicy::kSerial:
+            fSeqPool = std::unique_ptr<ROOT::TSequentialExecutor>(new ROOT::TSequentialExecutor());
+            break;
+#ifdef R__USE_IMT
+         case ROOT::Internal::ExecutionPolicy::kMultithread:
+            fThreadPool = std::unique_ptr<ROOT::TThreadExecutor>(new ROOT::TThreadExecutor(nProcessingUnits));
+            break;
+#endif
+         case ROOT::Internal::ExecutionPolicy::kMultiprocess:
+            fProcPool = std::unique_ptr<ROOT::TProcessExecutor>(new ROOT::TProcessExecutor(nProcessingUnits));
```

Comment:
what happens if `nProcessingUnits == 0` and `kMultiprocess`?

Reply:
It will choose the default number of workers (number of processors)

Reply:
is that potentially different from the default number of threads that IMT will use?

Reply:
It wasn't, but it is after the changes we introduced some months ago taking into account CPU bandwidth control. We didn't propagate them in TProcessExecutor. I will write down a note to open another PR.